torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -938,6 +946,8 @@ def __init__(
         self.training_state = TrainingState_.IDLE
         self.cpu_offload = cpu_offload or CPUOffload()
         self.backward_prefetch = backward_prefetch
+        self.limit_all_gathers = limit_all_gathers
+        self._max_num_inflight_all_gathers = 2  # empirically chosen
```

Comment:
a - I think it would be useful to expose "max_num_inflight_all_gathers" during init rather than embedding it, esp to allow larger customers to tune to their own hardware/models?  (I'm going to do this while testing this fix for example). 
b - If that is useful, then these two vars could also be compressed into a single 
self.max_num_all_gathers=2,
where 0 is Off, and thus no limit, and then 1, 2, etc. with default of 2, to simplify api and internal checks, rather than exposing two params?

Reply:
I had a previous version that did what you suggested (compress into one variable). I changed it since I thought it might be confusing for non-advanced users, for whom it may be better to expose a default value. For advanced users, we can expose a setter method that can be called post-construction. Of course, all of this is open for discussion. We are still thinking about the design.

I am not updating this PR at the moment as I experiment with different variations. I am planning to experiment with a size-based rule instead of a count-based rule and to flush the queue instead of only dequeuing a single event at a time. If your experiments are not urgent, I may recommend considering to wait a few days until I solidify the approach more.

Reply:
Sounds good - I'll hold off on doing much work with this fix then re: possible move to size based.  
Thanks for the updates!